torch/nn/parameter.py

```diff
@@ -42,3 +46,26 @@ def __reduce_ex__(self, proto):
             torch._utils._rebuild_parameter,
             (self.data, self.requires_grad, OrderedDict())
         )
+
+
+class _UninitializedParameter(Parameter):
```

Comment:
Do I read this correctly that the uninitialized param is an empty Tensor that will still be impacted by the to() calls and so we can reuse this info when creating the weight? Sounds great !

Reply:
I didn't think of it like that in the beginning, but after reading your comment, it just clicked-in and felt natural :).
I added a `materialize` method as per your other suggestion so now we can move around the module, or cast the parameter type before doing the actual initialization  :D